File: pingpong.c
#include <stdio.h>
#include <inttypes.h>
#include "coroutine.h"

void ping(coroutine_context_t ctx, void *arg);
void pong(coroutine_context_t ctx, void *arg);

int main(int argc, char **argv)
{
    coroutine_context_t ctx = coroutine_create_context(0, NULL);
    coroutine_main(ctx, (coroutine_func_t)ping, NULL);
    coroutine_free_context(ctx);
}

void ping(coroutine_context_t ctx, void *arg)
{
    uintptr_t n = 1;
    coroutine_t co_pong = coroutine_create(ctx, (coroutine_func_t)pong);
    do
    {
        printf("ping! %zi\n", n);
        n = (uintptr_t)coroutine_resume(ctx, co_pong, (void *)(n + 1));
    }
    while(coroutine_alive(co_pong));
    coroutine_free(co_pong);
}

void pong(coroutine_context_t ctx, void *arg)
{
    uintptr_t n = (uintptr_t)arg;
    while(n < 10)
    {
        printf("pong! %zi\n", n);
        n = (uintptr_t)coroutine_yield(ctx, (void *)(n + 1));
    }
}

File: examples/pingpong.lua
function ping(arg)
    local n = arg
    local ok = true
    local co_pong = coroutine.create(pong)
    repeat
        print("ping! " .. n)
        ok, n = coroutine.resume(co_pong, n + 1)
    until not ok or coroutine.status(co_pong) == 'dead'
end

function pong(arg)
    local n = arg
    while n < 10 do
        print("pong! " .. n)
        n = coroutine.yield(n + 1)
    end
end

ping(1)